From 74d50fa473e33520b541e9c17862834558caca8f Mon Sep 17 00:00:00 2001
From: syed touqeer pasha <quic_spasha@quicinc.com>
Date: Fri, 13 Sep 2024 11:41:51 +0530
Subject: [PATCH] QSDK: wifi: ath12k: Extract FST table from ag during SFE
 delete entry

As FST table is created and table context is attached to hw group.
During FST entry delete, get FST table context from hw group and
avoid vif dereference

Signed-off-by: syed touqeer pasha <quic_spasha@quicinc.com>
---
 drivers/net/wireless/ath/ath12k/fse.c | 30 +++++++++++++++++++++------
 drivers/net/wireless/ath/ath12k/fse.h |  2 +-
 drivers/net/wireless/ath/ath_fse.c    | 26 +++--------------------
 3 files changed, 28 insertions(+), 30 deletions(-)

diff --git a/drivers/net/wireless/ath/ath12k/fse.c b/drivers/net/wireless/ath/ath12k/fse.c
index 314fc39..ed7f1b9 100644
--- a/drivers/net/wireless/ath/ath12k/fse.c
+++ b/drivers/net/wireless/ath/ath12k/fse.c
@@ -75,7 +75,7 @@ void *ath12k_fse_get_ab_from_vif(struct ieee80211_vif *vif,
 	return ab;
 }
 
-static void ath12k_dp_rx_sfe_update_flow_info(struct ath12k_base *ab,
+static void ath12k_hw_grp_dp_rx_sfe_update_flow_info(struct ath12k_hw_group *ag,
 					      struct rx_flow_info *flow_info,
 					      u32 *src_ip, u32 src_port,
 					      u32 *dest_ip, u32 dest_port,
@@ -83,7 +83,7 @@ static void ath12k_dp_rx_sfe_update_flow_info(struct ath12k_base *ab,
 {
 	struct hal_flow_tuple_info *tuple_info = &flow_info->flow_tuple_info;
 
-	ath12k_dbg(ab, ATH12K_DBG_DP_FST, "%s S_IP:%x:%x:%x:%x,sPort:%u,D_IP:%x:%x:%x:%x,dPort:%u,Proto:%d,Ver:%d",
+	ath12k_dbg(ag->ab[0], ATH12K_DBG_DP_FST, "%s S_IP:%x:%x:%x:%x,sPort:%u,D_IP:%x:%x:%x:%x,dPort:%u,Proto:%d,Ver:%d",
 		   fse_state_to_string(operation),
 		   src_ip[0], src_ip[1], src_ip[2], src_ip[3], src_port, dest_ip[0],
 		   dest_ip[1], dest_ip[2], dest_ip[3], dest_port, protocol, version);
@@ -121,24 +121,42 @@ int ath12k_sfe_add_flow_entry(struct ath12k_base *ab,
 	if (!ath12k_fse_enable)
 		return -EINVAL;
 
-	ath12k_dp_rx_sfe_update_flow_info(ab, &flow_info, src_ip, src_port, dest_ip,
+	ath12k_hw_grp_dp_rx_sfe_update_flow_info(ab->ag, &flow_info, src_ip, src_port, dest_ip,
 					  dest_port, protocol, version, FSE_RULE_ADD);
 
 	return ath12k_dp_rx_flow_add_entry(ab, &flow_info);
 }
 
-int ath12k_sfe_delete_flow_entry(struct ath12k_base *ab,
+int ath12k_sfe_delete_flow_entry(struct ieee80211_hw *hw,
 				 u32 *src_ip, u32 src_port,
 				 u32 *dest_ip, u32 dest_port,
 				 u8 protocol, u8 version)
 {
 	struct rx_flow_info flow_info = {0};
+	struct ath12k_hw *ah = NULL;
+	struct ath12k_hw_group *ag = NULL;
 
 	if (!ath12k_fse_enable)
 		return -EINVAL;
 
-	ath12k_dp_rx_sfe_update_flow_info(ab, &flow_info, src_ip, src_port, dest_ip,
+	ah = hw->priv;
+	if (!ah) {
+		ath12k_err("HW invalid-Flow delete failed:S_IP:%x:%x:%x:%x,sPort:%u,D_IP:%x:%x:%x:%x,dPort:%u,Proto:%d,Ver:%d",
+			   src_ip[0], src_ip[1], src_ip[2], src_ip[3], src_port, dest_ip[0],
+			   dest_ip[1], dest_ip[2], dest_ip[3], dest_port, protocol, version);
+		return -EINVAL;
+	}
+
+	ag = ah->ag;
+	if (!ag) {
+		ath12k_err("HW grp invalid-Flow delete failed:S_IP:%x:%x:%x:%x,sPort:%u,D_IP:%x:%x:%x:%x,dPort:%u,Proto:%d,Ver:%d",
+			   src_ip[0], src_ip[1], src_ip[2], src_ip[3], src_port, dest_ip[0],
+			   dest_ip[1], dest_ip[2], dest_ip[3], dest_port, protocol, version);
+		return -EINVAL;
+	}
+
+	ath12k_hw_grp_dp_rx_sfe_update_flow_info(ag, &flow_info, src_ip, src_port, dest_ip,
 					  dest_port, protocol, version, FSE_RULE_DELETE);
 
-	return ath12k_hw_grp_dp_rx_flow_delete_entry(ab->ag, &flow_info);
+	return ath12k_hw_grp_dp_rx_flow_delete_entry(ag, &flow_info);
 }
diff --git a/drivers/net/wireless/ath/ath12k/fse.h b/drivers/net/wireless/ath/ath12k/fse.h
index aa230d3..4fda24e 100644
--- a/drivers/net/wireless/ath/ath12k/fse.h
+++ b/drivers/net/wireless/ath/ath12k/fse.h
@@ -40,7 +40,7 @@ int ath12k_sfe_add_flow_entry(struct ath12k_base *ab,
 			      u32 *src_ip, u32 src_port,
 			      u32 *dest_ip, u32 dest_port,
 			      u8 protocol, u8 version);
-int ath12k_sfe_delete_flow_entry(struct ath12k_base *ab,
+int ath12k_sfe_delete_flow_entry(struct ieee80211_hw *hw,
 				 u32 *src_ip, u32 src_port,
 				 u32 *dest_ip, u32 dest_port,
 				 u8 protocol, u8 version);
diff --git a/drivers/net/wireless/ath/ath_fse.c b/drivers/net/wireless/ath/ath_fse.c
index 90dcf63..89f20a1 100644
--- a/drivers/net/wireless/ath/ath_fse.c
+++ b/drivers/net/wireless/ath/ath_fse.c
@@ -106,7 +106,7 @@ bool ath_fse_add_rule(struct ath_fse_flow_info *fse_info)
 			 * the other direction as well.
 			 */
 			if (src_wdev) {
-				fse_cb->fse_rule_delete(ab,
+				fse_cb->fse_rule_delete(hw,
 						fse_info->src_ip, fse_info->src_port,
 						fse_info->dest_ip, fse_info->dest_port,
 						fse_info->protocol, fse_info->version);
@@ -123,9 +123,7 @@ EXPORT_SYMBOL(ath_fse_add_rule);
 bool ath_fse_delete_rule(struct ath_fse_flow_info *fse_info)
 {
 	struct wireless_dev *src_wdev, *dest_wdev;
-	struct ieee80211_vif *vif;
 	struct ieee80211_hw *hw;
-	void *ab;
 	int fw_ret = 0;
 	int rv_ret = 0;
 
@@ -156,16 +154,7 @@ bool ath_fse_delete_rule(struct ath_fse_flow_info *fse_info)
 		if (!ieee80211_hw_check(hw, SUPPORT_ECM_REGISTRATION))
 			return false;
 
-		vif = wdev_to_ieee80211_vif_vlan(src_wdev, false);
-		if (!vif)
-			return false;
-
-		ab = fse_cb->fse_get_ab(vif, fse_info->src_mac);
-		if (!ab) {
-			pr_debug("Failed to delete a rule in FST<ab NULL>");
-			return false;
-		}
-		fw_ret = fse_cb->fse_rule_delete(ab,
+		fw_ret = fse_cb->fse_rule_delete(hw,
 						  fse_info->src_ip, fse_info->src_port,
 						  fse_info->dest_ip, fse_info->dest_port,
 						  fse_info->protocol, fse_info->version);
@@ -176,16 +165,7 @@ bool ath_fse_delete_rule(struct ath_fse_flow_info *fse_info)
 		if (!ieee80211_hw_check(hw, SUPPORT_ECM_REGISTRATION))
 			return false;
 
-		vif = wdev_to_ieee80211_vif_vlan(dest_wdev, false);
-		if (!vif)
-			return false;
-
-		ab = fse_cb->fse_get_ab(vif, fse_info->dest_mac);
-		if (!ab) {
-			pr_debug("Failed to delete a rule in FST<ab NULL>");
-			return false;
-		}
-		rv_ret = fse_cb->fse_rule_delete(ab,
+		rv_ret = fse_cb->fse_rule_delete(hw,
 						  fse_info->dest_ip, fse_info->dest_port,
 						  fse_info->src_ip, fse_info->src_port,
 						  fse_info->protocol, fse_info->version);
-- 
2.34.1

